mathics/algorithm/introselect.py

```diff
@@ -0,0 +1,170 @@
+    # in [i, j]. this is not strictly necessary, but it might pinpoint the
+    # sort position of one more element at this point, which is good.
+
+    if f > i:
+        a[i], a[f] = a[f], a[i]
+        i += 1
+    elif f < j:
+        a[f], a[j] = a[j], a[f]
+        j -= 1
+
+    # leaving here, we know three things (see Hoare's original description):
+
+    # (1) a[r] <= pivot for r <= j
+    # (2) a[r] = pivot for j < r < i
+    # (3) a[r] >= pivot for r >= i
+
+    # note that the elements a[r] for j < r < i have already reached their
+    # final sorting position in the array, i.e. if we look for the location
+    # of one of these, we're done.
+
+    return j, i
+
+
+def bfprt(a, k):  # changes a
+    while True:
+        if len(a) == 1:
+            return a[0]
+
+        median_indices = [_median5(a[i:i + 5]) for i in range(0, len(a), 5)]
+        median_values = [a[5 * b + i] for b, i in enumerate(median_indices)]
+        pivot = bfprt(median_values[:], len(median_values) // 2)
+        pivot_block_index = [id(x) for x in median_values].index(id(pivot))
+        pivot_index = 5 * pivot_block_index + median_indices[pivot_block_index]
+
+        i, j = _partition(a, pivot_index)
+        if k <= i:
+            a = a[:i + 1]
+        elif k >= j:
+            a = a[j:]
+            k -= j
+        else:
+            return a[k]
+
+
+def introselect(a, k):  # changes a
+    depth = len(a).bit_length() * 2  # see algorithm INTROSORT in [2], page 5
+
+    while len(a) >= 3 and depth > 0:
+        p = _median3(a)
+
+        # see Algorithm 65 FIND in [1]
+        i, j = _partition(a, p)
+        if k <= i:
+            a = a[:i + 1]
+        elif k >= j:
+            a = a[j:]
+            k -= j
+        else:
+            return a[k]
+
+        depth -= 1
+
+    if depth <= 0 and len(a) > 3:
+        return bfprt(a, k)
+    else:
+        return sorted(a)[k]
+
+
+if __name__ == "__main__":
+    import random
+
+    def test_algorithm(l, r_max, name, f):
+        a = [random.randint(-r_max, r_max) for _ in range(l)]
+        b = sorted(a)
+        c = [f(a[:], i) for i in range(len(a))]
+        if b == c:
+            print('OK %s r: %d l: %d' % (name, r_max, l))
+            return True
+        else:
+            print('FAIL %s r: %d l: %d' % (name, r_max, l))
+            print(a, b, c)
+            return False
+
+    def test_configuration(l_max, r_max):
+        for l in range(l_max):
+            if not test_algorithm(l, r_max, 'bfprt', bfprt):
+                return False
+
+        for l in range(l_max):
+            if not test_algorithm(l, r_max, 'introselect', introselect):
+                return False
+
+        return True
+
+    def test_range(l_max):
+        # we test two cases: many same elements, and few same elements.
+        return test_configuration(l_max, 10) and test_configuration(200, 1000)
+
+    if test_range(200):
+        print('ALL OK.')
+    else:
+        print('ABORTED WITH FAILURE.')
```

Comment:
I think a better place for these tests would be e.g. `test/algorithm/test_introselect.py`.